doc/man7/openssl-core.h.pod

```diff
@@ -0,0 +1,200 @@
+pointers).
+
+These arrays can be used both to pass parameter data to some function,
+but can also be used to request data from some function.
+For a request, this description talks about a requesting function that
+provides the C<OSSL_PARAM> array, and a responding function that fills
+in the buffers that the C<OSSL_PARAM> elements point at.
+
+Regardless of the use, the array as well as any buffer space it points
+to must be provided by the caller (the requesting function).
+
+The fields of this structure are:
+
+=over 4
+
+=item C<key>
+
+The identity of the parameter in the form of a string.
+
+These identities are defined by the entity that provides these
+parameters, i.e. names for parameters provided by the OpenSSL
+libraries are defined by the libraries, and names for parameters
+provided by providers are defined by those providers, except for the
+pointer form of strings (see data type descriptions below)
+
+=item C<buffer>
+
+=item C<buffer_size>
+
+C<buffer> is a pointer to the memory where the parameter data is or
+shall be stored, and C<buffer_size> is its size in bytes.
+The organization of the data depends on the parameter type and flag.
+
+=item C<return_size>
+
+When an array of C<OSSL_PARAM> is used to request data, the responding
+function must set this field to indicate the actual size of the data
+stored in memory pointed at by the C<buffer> pointer, or the needed
+buffer size if the size indicated by C<buffer_size> is too small.
+
+=item C<data_type>
+
+=for comment It's still debated if this field should be present, or if
+the type should always be implied by how it's used.  Either way, these
+data types will have to be passed together with the names as an array
+of OSSL_ITEM, for discovery purposes.
+
+The C<data_type> is a value that describes the type and organization of
+the data, and can be one of:
+
+=over 4
+
+=item C<OSSL_PARAM_INTEGER>
+
+=item C<OSSL_PARAM_UNSIGNED_INTEGER>
+
+The parameter data is an integer (signed or unsigned) of arbitrary
+length, organized in native form, i.e. most significant byte first on
+Big-Endian systems, and least significant byte first on Little-Endian
+systems.
+
+=item C<OSSL_PARAM_REAL>
+
+=for comment It's still debated if we need this or not.
+
+The parameter data is a floating point value in native form.
+
+=item C<OSSL_PARAM_UTF8_STRING>
+
+The parameter data is a printable string.
+
+The pointer form
+
+=item C<OSSL_PARAM_OCTET_STRING>
+
+The parameter data is an arbitrary string of bytes.
+
+=back
+
+Additionally, this flag can be added to any type:
+
+=over 4
+
+=item C<OSSL_PARAM_POINTER_FLAG>
+
+This can be used to indicate that constant data is or will be passed.
+It works by assuming that the memory pointed at by the C<buffer> field
+contains a pointer to the actual data.
+C<buffer_size> must be set to the size of the data pointed at rather
+than the size of the memory pointed at by the C<buffer> field.
```

Comment:
I'm not sure I understand the distinction between these two cases, i.e. what's the difference between the size of the "data" pointed at vs the size of the "memory" pointed at.

Reply:
Does this explain it?

``` C
const char *foo = NULL;
size_t foo_l;
char bar[1024];
size_t bar_l;
const OSSL_PARAM request[] = {
    { "foo", OSSL_PARAM_UTF8_STRING_PTR, &foo, sizeof(foo), &foo_l },
    { "bar", OSSL_PARAM_UTF8_STRING, &bar, sizeof(bar), &bar_l },
    { NULL, 0, NULL, 0, NULL }
};
```

A responding function would typically do this (highly simplified, of course):

``` C
/* const OSSL_PARAM *params */
*(void **)params[0].buffer = "foo value";
*params[0].return_size = 10; /* size of "foo value" incl NUL */
memcpy(params[1].buffer, "bar value", 10);
*params[1].return_size = 10; /* size of "bar value" incl NUL */
```

Reply:
The "memory" that the `buffer` field is pointing at is `foo` and `bar`.

Maybe you can come up with a better phrasing...  It does feel clumsy, so any help appreciated

Reply:
So in one case we're talking about the size of the pointer itself while in the other we are talking about the size of the memory pointed at by the pointer?

Reply:
Argh, I'm tired...  No, that doesn't match what I intended.  Actually, for a *request*, `buffer_size` isn't at all relevant.  However, for just *setting* parameters, `buffer_size` should be the size of the *data*, not the size of the pointer.  In other words, the `request` array above should have looked like this:

``` C
const OSSL_PARAM request[] = {
    { "foo", OSSL_PARAM_UTF8_STRING_PTR, &foo, 0 /*irrelevant*/, &foo_l },
    { "bar", OSSL_PARAM_UTF8_STRING, &bar, sizeof(bar), &bar_l },
    { NULL, 0, NULL, 0, NULL }
};
```

However, if you want to *set* params on some object, it would look like this:

``` C
const char *foo = "some string";
const char bar[] = "some other string";
const OSSL_PARAM set[] = {
    { "foo", OSSL_PARAM_UTF8_STRING_PTR, &foo, strlen(foo), NULL },
    { "bar", OSSL_PARAM_UTF8_STRING, &bar, sizeof(bar), NULL },
    { NULL, 0, NULL, 0, NULL }
};
```

I'll rewrite that piece of text.

Reply:
I'm starting to wonder if `OSSL_PARAM` shouldn't have a manpage of its own.  Question is what section it should be placed in (section 3 manpages are typically function oriented, types are not normally described on their own, and are usually not mentioned among the NAMES...  we have cheated in actually naming them in some documentation, but it's still been together with functions that use them...  can we get away with placing a manpage that's *only* about a type in section 3?)

Reply:
I've rephrased and provided a few examples, does that clarify things?

Reply:
We'll end up with some utility functions to access PARAMs, so a section 3 man page would be appropriate.

Reply:
e.g. a helper function to deal with the differences between the STRING and STRING_PTR types.  Fundamentally, they are for doing the same thing.